iroh/src/node.rs

```diff
@@ -152,27 +153,33 @@ impl<D: BaoStore> Node<D> {
         self.inner.endpoint.home_relay()
     }
 
-    /// Aborts the node.
+    /// Shutdown the node.
     ///
     /// This does not gracefully terminate currently: all connections are closed and
-    /// anything in-transit is lost.  The task will stop running.
-    /// If this is the last copy of the `Node`, this will finish once the task is
-    /// fully shutdown.
+    /// anything in-transit is lost. The shutdown behaviour will become more graceful
+    /// in the future.
     ///
-    /// The shutdown behaviour will become more graceful in the future.
+    /// Returns a future that completes once all tasks terminated and all resources are closed.
+    /// The future resolves to an error if the main task panicked.
     pub async fn shutdown(self) -> Result<()> {
+        // Trigger shutdown of the main run task by activating the cancel token.
         self.inner.cancel_token.cancel();
 
-        if let Ok(task) = Arc::try_unwrap(self.task) {
-            task.await?;
-        }
+        // Wait for the main task to terminate.
+        self.task.await.map_err(|err| anyhow!(err))?;
+
         Ok(())
     }
 
     /// Returns a token that can be used to cancel the node.
     pub fn cancel_token(&self) -> CancellationToken {
         self.inner.cancel_token.clone()
     }
+
+    /// Get a protocol handler.
+    pub fn get_protocol<P: Protocol>(&self, alpn: &[u8]) -> Option<Arc<P>> {
```

Comment:
Not for this PR, but I was wondering if we should have a generic "place to put stuff" on the node that also works for non protocols. E.g. the blobs db or the docs db or the downloader, stuff that you might want to use from multiple protocols but that is not a protocol.

Reply:
Yup, let's discuss this separately though. Right now the unspawned node / protocol builder gives access to *more* stuff than the final node. But we also want to keep the API of the node small and streamlined, and I think generally advise to work with the client if possible to not have to increase the public API surface too much. Protocol handlers are the kind of "exception" IMO so I think it's fine to have more things on the unspawned node than on the final node (and everything is clone I think).